File: python/triangle/triangle.py
def triangle(sides):
    sides.sort()

    return sides[0] > 0 and sides[0] + sides[1] >= sides[2]


def equilateral(sides):
    sides.sort()

    return triangle(sides) and sides[0] == sides[2]


def isosceles(sides):
    sides.sort()

    return triangle(sides) and (sides[0] == sides[1] or sides[1] == sides[2])


def scalene(sides):
    sides.sort()

    return triangle(sides) and sides[0] != sides[1] and sides[1] != sides[2]


def degenerate(sides):
    sides.sort()

    return triangle(sides) and sides[0] + sides[1] == sides[2]

File: haskell/triangle/src/Triangle.hs
module Triangle
  ( TriangleType(..)
  , triangleType
  ) where

import           Data.List                      ( sort )

data TriangleType = Equilateral
                  | Isosceles
                  | Scalene
                  | Degenerate
                  | Illegal
                  deriving (Eq, Show)

triangleType :: (Num a, Ord a) => a -> a -> a -> TriangleType
triangleType a b c | x <= 0 || x + y < z = Illegal
                   | x == z              = Equilateral
                   | x == y || y == z    = Isosceles
                   | x + y == z          = Degenerate
                   | otherwise           = Scalene
  where x : y : z : _ = sort [a, b, c]
